src/main/java/com/google/devtools/build/lib/rules/python/PyRuntime.java

```diff
@@ -68,15 +67,23 @@ public ConfiguredTarget create(RuleContext ruleContext)
 
     Artifact coverageTool = null;
     NestedSet<Artifact> coverageFiles = null;
+    TransitiveInfoCollection coverageTarget = ruleContext.getPrerequisite("coverage_tool");
     if (coverageTarget != null) {
-      FilesToRunProvider filesToRun = coverageTarget.getProvider(FilesToRunProvider.class);
-      if (filesToRun == null) {
-        ruleContext.attributeError("coverage_tool", "must be an executable target.");
+      NestedSet<Artifact> coverageToolFiles =
+          coverageTarget.getProvider(FileProvider.class).getFilesToBuild();
+      if (coverageToolFiles.isSingleton()) {
+        coverageTool = coverageToolFiles.getSingleton();
       } else {
-        coverageTool = filesToRun.getExecutable();
+        FilesToRunProvider filesToRun = coverageTarget.getProvider(FilesToRunProvider.class);
```

Comment:
Perhaps more generally applicable, requiring `PyInfo` and a single file, without allowing an executable target, would rule out using a `py_binary` target here, as that doesn't produce a `PyInfo` and its output contains at least two files (including the stub).

Reply:
I appreciate the honesty =).

-----

re: using a py_binary here: I don't think that's even possible -- a py_binary depends on the toolchain, so putting a py_binary as a dependency of the toolchain will result in circular dependencies.

(aside: I think there's a variety of current or potential circular dependency/bootstrapping issues we need to figure out; let's take that convo outside this PR though)

re: running `coverage run main.py` instead of `python coverage.py run main.py`: I don't think it's as far fetched as you might think :). But let's dive into that outside this PR; it's enabling that sort of thing I'd like to bring to the Bazel Python rules.

-----

I'm willing to accept an executable here because I can appreciate your situation, and as long as we make sure that (via docs, in this case), whatever executable is put here, it abides by the conceptual contract of "(1) its executable output is a Python entry point (e.g. .py file), not an e.g. application, and (2) the binary is compatible with being treated as a dependency of other binaries and merged in as such".

I'll take a look at the doc here tonight and suggest changes, if any are needed.

That said, I still want get across an important point about Bazel design and how targets interact with one another. Sorry that I've pontificated a bunch below; I'm sincerely just interested in moving the Python rules forward and making sure they have a solid design base we can build upon.

-----

re: Allowing an executable for the coverage dependency:
First, to make sure we're on the same page, when I say "binary" I mean a top-level executable problem. I don't mean e.g. a shared library. I say this because something I've seen in OSS+Bazel+Python is, e.g. using `cc_binary` to produce a Python-loadable shared-library, and I've seen more C-centric people call them a binary.

Second, the way I see things, when we build e.g. a `py_binary`, things like the Python interpreter/runtime and coverage are inputs to that binary. This is because, essentially, that's how binary rules in Bazel are implemented: the binary rule gathers pieces from the toolchain, dependencies, etc and puts it all together into a final executable. But also because binaries, as the top-level entry point, should get the final say in how they work. (To be fair, I probably have some bias because, within Google, we use embedded Python, so Python is more a library used within a C program than separate runtime).

This is distinct from another way of thinking of things: that the `py_binary` and coverage are inputs to the Python interpreter (or that the py binary and interpreter are inputs to the coverage tool). This is more conceptually how non-Bazel Python works (you install a Python runtime, the add stuff into that runtime).

Insofar as Bazel is concerned, mixing together executables is problematic. In this case, we're mixing together a coverage executable and the py_binary executable. There are several ways this can interact badly, but it all comes down to each target specifying something that is mutually exclusive with something else. Examples include:
  * Specifying different malloc implementations (in Google, we allow `py_binary` to specify a malloc implementation)
  * Specifying or having different versions of the same the same library
  * Having dependencies that are mutually exclusive (e.g. I think boringssl/openssl are mutually exclusive; internally we have other such cases).
  * etc

A similar problem can happen at the Starlark-provider metadata level: an inner binary's settings (e.g. C++ info) can propagate up to an outer binary. This is one of the reasons `py_binary` _doesn't_ provide PyInfo. We saw this sort of issue internally very frequently when doing our Python 2 to 3 transition (a python 2 binary in the dependencies of a Python 3 binary, or vice versa, would fubar the build).

A couple other problems also occur when a binary is added to the dependencies (i.e. `deps` or equiv; not `data`) of another binary: build bloat and unnecessary overhead. This comes in a couple of different forms.

Using `py_binary` as an example again, recall that it (currently) has two outputs: the executable stub script and main .py file. We actually don't care about that executable stub script in this case, but it still has to be built and included into the runfiles (we're merging in its runfiles, as we must). If we built for e.g. windows, then we might get a `.exe` launcher that is entirely unused. In today's rules, this isn't too much overhead (the stub script is small), so nobody really notices. With something like embedded Python (or e.g. pyoxider etc), that overhead is pretty large.

This also segues into the next issue: the inner binary respects the flag settings intended for the outer binary. This is because it's being built in the same configuration (target, as it must). Going back to `py_binary`, if you specified, I think its the `--build_zip` flag (sp), then the inner binary would also build a (largely self-contained) zip file and use that as its output instead of e.g. the typical stub script. This isn't uncommon for binaries. While libraries can also have flags, it's not really an issue -- they still abide by the "I'm a thing to be included in another thing" contract.

I use `py_binary` as a convenient, if presently impossible, example. The same could be true of e.g. `sh_binary` etc.

Reply:
> re: using a py_binary here: I don't think that's even possible -- a py_binary depends on the toolchain, so putting a py_binary as a dependency of the toolchain will result in circular dependencies.

Yes, you're absolutely right, and I'd thought of that about half an hour after making the comment.  But there are other similar rules which take in some script and output a version of that script with some kind of wrapper or whatever.  The point, here, is that one can imagine a number of different rules which output more than one file, and that's fine so long as the rule designates no more than one file as the entry point.

> I'm willing to accept an executable here because I can appreciate your situation, and as long as we make sure that (via docs, in this case), whatever executable is put here, it abides by the conceptual contract of "(1) its executable output is a Python entry point (e.g. .py file), not an e.g. application, and (2) the binary is compatible with being treated as a dependency of other binaries and merged in as such".
>
> I'll take a look at the doc here tonight and suggest changes, if any are needed. 

That's very reasonable.  I'll grant the docstring could use some expansion; I didn't expand on it much between my initial draft and now.

> That said, I still want get across an important point about Bazel design and how targets interact with one another. Sorry that I've pontificated a bunch below; I'm sincerely just interested in moving the Python rules forward and making sure they have a solid design base we can build upon.
> 
> re: Allowing an executable for the coverage dependency: First, to make sure we're on the same page, when I say "binary" I mean a top-level executable problem. I don't mean e.g. a shared library. I say this because something I've seen in OSS+Bazel+Python is, e.g. using `cc_binary` to produce a Python-loadable shared-library, and I've seen more C-centric people call them a binary.

Yes, one doesn't have to spend very much time writing (starlark) bazel rules to get that beaten into you :-).

> Second, the way I see things, when we build e.g. a `py_binary`, things like the Python interpreter/runtime and coverage are inputs to that binary. This is because, essentially, that's how binary rules in Bazel are implemented: the binary rule gathers pieces from the toolchain, dependencies, etc and puts it all together into a final executable. But also because binaries, as the top-level entry point, should get the final say in how they work. (To be fair, I probably have some bias because, within Google, we use embedded Python, so Python is more a library used within a C program than separate runtime).
> 
> This is distinct from another way of thinking of things: that the `py_binary` and coverage are inputs to the Python interpreter (or that the py binary and interpreter are inputs to the coverage tool). This is more conceptually how non-Bazel Python works (you install a Python runtime, the add stuff into that runtime).

In this case it's actually that the interpreter and the coverage tool are inputs to `py_binary`.  The `py_binary` stub (which could very reasonably not be written in python at all; making it a small C executable for example would solve more than [one problem](https://github.com/bazelbuild/bazel/issues/8446), and somewhat ironically, probably make things more portable) has the job of gluing those pieces together into some arraignment that produces the desired result: to run the given python script, with the given python interpreter, optionally using the given tool to collect coverage along the way.

> Insofar as Bazel is concerned, mixing together executables is problematic. In this case, we're mixing together a coverage executable and the py_binary executable. There are several ways this can interact badly, but it all comes down to each target specifying something that is mutually exclusive with something else.

Yes, and it's pretty clear in any case, from the implementation here, that the coverage tool does need to be something that  python can interpret (though as you mention, that could be an extension module or a `.pyc` bytecode file).  Even beyond that, it specifically expects the command line API to conform to that of [`coveragepy`](https://coverage.readthedocs.io/); it would be pretty far-fetched to call the implementation generic.  That said, I do believe that this is what ~99% of users outside of google would want anyway.  There aren't many organizations out there with enough SWE capacity to justify reinventing that particular wheel.

Perhaps using `FilesToRun` for the purpose of designating an entry point for a rule that outputs several files, when the target isn't _really_ a standalone binary, is a bit of an abuse of the mechanism, but if so I at least consider it to be a small one.  In our rule set at least, `bazel run .../coverage_tool`, _will_ work, though it'll run it through the system python interpreter, which is not ideal.

> * Specifying different malloc implementations (in Google, we allow `py_binary` to specify a malloc implementation)
> * Specifying or having different versions of the same the same library
> * Having dependencies that are mutually exclusive (e.g. I think boringssl/openssl are mutually exclusive; internally we have other such cases).
> * etc

Again, while it's true the API here is specified as if it were, we're not actually talking about arbitrary tools here.  Coveragepy _does_ include extension modules, however, which could bring up "interesting" issues with e.g. malloc implementations; I don't think there's really much of a way around that; if you're building `coveragepy` from source within bazel, which is probably a better idea that just using the precompiled wheel, then at least it's _likely_ that it'll be built with compatible options.

> A similar problem can happen at the Starlark-provider metadata level: an inner binary's settings (e.g. C++ info) can propagate up to an outer binary. This is one of the reasons `py_binary` _doesn't_ provide PyInfo. We saw this sort of issue internally very frequently when doing our Python 2 to 3 transition (a python 2 binary in the dependencies of a Python 3 binary, or vice versa, would fubar the build).

I think this issue is largely sidestepped here because we're not actually adding anything to the explicit dependencies of the `py_bianry` or `py_test` rule that's getting coverage enabled - it's only through the implicit toolchain dependency, and then only when coverage is enabled.  Meanwhile the toolchain provider metadata does _not_ propagate any metadata from the coverage tool, and as far as I know aspects don't follow implicit (toolchain) dependencies so there's not really any way for that to be visible to a starlark rule.

> A couple other problems also occur when a binary is added to the dependencies (i.e. `deps` or equiv; not `data`) of another binary: build bloat and unnecessary overhead. This comes in a couple of different forms.
> 
> Using `py_binary` as an example again, recall that it (currently) has two outputs: the executable stub script and main .py file. We actually don't care about that executable stub script in this case, but it still has to be built and included into the runfiles (we're merging in its runfiles, as we must). If we built for e.g. windows, then we might get a `.exe` launcher that is entirely unused. In today's rules, this isn't too much overhead (the stub script is small), so nobody really notices. With something like embedded Python (or e.g. pyoxider etc), that overhead is pretty large.

I can't really imagine many situations where someone would want to make bundled distributions of a tool (e.g. pyoxidizer) with coverage enabled.  Maybe for python code that needed to run on some specific hardware?  But maybe that's just my imagination being too limited.  At any rate if you did that you'd be having a bad time because that `.exe` entry point is what the stub would be feeding to the python interpreter, which probably wouldn't go very well.

> This also segues into the next issue: the inner binary respects the flag settings intended for the outer binary. This is because it's being built in the same configuration (target, as it must). Going back to `py_binary`, if you specified, I think its the `--build_zip` flag (sp), then the inner binary would also build a (largely self-contained) zip file and use that as its output instead of e.g. the typical stub script. This isn't uncommon for binaries. While libraries can also have flags, it's not really an issue -- they still abide by the "I'm a thing to be included in another thing" contract.
> 
> I use `py_binary` as a convenient, if presently impossible, example. The same could be true of e.g. `sh_binary` etc.

Thank you, these are fascinating insights which help to crystalize some things about binary targets in bazel which were, admittedly, fuzzy in my understanding.  That said, I don't really think they apply here, for reasons already discussed above.

Reply:
Our team actually hasn't had any trouble integrating extension modules into our bazel build, though we do need to depend explicitly on our python build target.  It would be pretty cool if python toolchains provided `CcInfo` for those purposes, to make it a bit more portable.  (don't want to create noise there by linking it but issue 24665 in grpc is somewhat related to this).

Yes, you are correct that `pytest` has its own coverage API (or maybe plugin? Not quite sure), however that doesn't support reporting in `lcov` format, which makes it less useful in a bazel context.  Agree it would be neat to add some kind of provider to make coverage more configurable, but looking at the implementation in `python_stub_template.txt` there's a lot in there that would be difficult to make configurable.

That functionality could be added at some point in the future without breaking backwards compatibility, with `PyRuntimeRule` detecting if such an provider was present and responding accordingly, but otherwise falling back on the behavior implemented here.